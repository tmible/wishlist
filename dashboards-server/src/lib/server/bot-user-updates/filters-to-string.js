/**
 * @typedef {
 *   import('$lib/bot-user-updates/initialization.js').BotUserUpdatesFilters
 * } BotUserUpdatesFilters
 */

/**
 * Преобразование объекта фильтров к строке для использования в SQL выражении
 * @function filtersToString
 * @param {BotUserUpdatesFilters} filters Фильтры
 * @returns {string} SQL представление фильтров
 */
export const filtersToString = (filters) => {
  const filtersString = Object.entries(filters)
    .map(([ key, value ]) => {
      if (typeof value === 'object') {
        const filter = [];
        if (value.start) {
          filter.push(`${key} >= ${value.start}`);
        }
        if (value.end) {
          filter.push(`${key} <= ${value.end}`);
        }
        return filter.join(' AND ');
      }
      if (value === '') {
        return '';
      }
      return `${key} = ${value}`;
    })
    .filter(Boolean)
    .join(' AND ');
  return `${filtersString.length > 0 ? ' AND ' : ''}${filtersString}`;
};
