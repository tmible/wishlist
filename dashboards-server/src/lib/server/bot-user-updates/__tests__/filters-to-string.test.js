import { describe, expect, it } from 'vitest';
import { filtersToString } from '../filters-to-string.js';

describe('bot user updates / filtersToString', () => {
  it('should return empty string', () => {
    expect(filtersToString({})).toBe('');
  });

  it('should return filters string', () => {
    expect(
      filtersToString({
        simple1: 'filter',
        simple2: '',
        range1: { start: 'start', end: 'end', random: 'property' },
        range2: { start: 'only' },
        range3: { end: 'only' },
        range4: {},
      }),
    ).toMatchSnapshot();
  });
});
