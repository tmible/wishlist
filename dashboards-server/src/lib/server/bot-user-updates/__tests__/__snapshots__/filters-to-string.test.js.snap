// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bot user updates / filtersToString > should return filters string 1`] = `" AND simple1 = filter AND range1 >= start AND range1 <= end AND range2 >= only AND range3 <= only"`;
